clusterloader2/api/validation.go

```diff
@@ -18,26 +18,187 @@ package api
 
 import (
 	"fmt"
+	"os"
 
+	"k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
 	"k8s.io/perf-tests/clusterloader2/pkg/errors"
 )
 
+// ConfigValidator contains metadata for config validation.
+type ConfigValidator struct {
+	basePath   string
+	testConfig *Config
+}
+
+// NewConfigValidator creates a new ConfigValidator object
+func NewConfigValidator(basePath string, testConfig *Config) *ConfigValidator {
+	return &ConfigValidator{
+		basePath:   basePath,
+		testConfig: testConfig,
+	}
+}
+
 // Validate checks and verifies the configuration parameters.
-func (conf *Config) Validate() *errors.ErrorList {
-	// TODO(#1636): Validate other aspects of the api
-	return conf.Namespace.Validate()
+func (v *ConfigValidator) Validate() *errors.ErrorList {
+	allErrs := field.ErrorList{}
+	c := v.testConfig
+
+	// TODO(#1696): Clean up after removing automanagedNamespaces
+	if c.AutomanagedNamespaces != 0 && c.AutomanagedNamespaces < 0 {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("automanagedNamespaces"), c.AutomanagedNamespaces, "must be non-negative"))
+	}
+
+	allErrs = append(allErrs, v.verifyNamespace(&c.Namespace, field.NewPath("namespace"))...)
+
+	if len(c.TuningSets) > 0 {
+		for i := range c.TuningSets {
+			allErrs = append(allErrs, v.verifyTuningSet(c.TuningSets[i], field.NewPath("tuningSets").Index(i))...)
+		}
+	}
+
+	if len(c.Steps) > 0 {
+		for i := range c.Steps {
+			allErrs = append(allErrs, v.verifyStep(c.Steps[i], field.NewPath("steps").Index(i))...)
+		}
+	}
+
+	if len(allErrs) == 0 {
+		return nil
+	}
+
+	return v.toErrorList(allErrs)
```

Comment:
let's just return `&allErrs`

[and remove toErrorList function compltely)]

Reply:
This was so that I could adhere to the `*errors.ErrorList` return value for [CompileTestConfig](https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/test/test.go#L68). I thought of converting it here because passing `*field.ErrorList` up would look out of place in test.go and clusterloader.go

Reply:
Hmm I could actually, but that would mean I have to do that same error conversion over here instead: https://github.com/kubernetes/perf-tests/pull/1694/files#diff-48e20ddd5ec57a1a2dcc78e5017d4b44137f0e94da165423c4ad333d47f967a5R105 in order to return `*errors.ErrorList` to the caller.  Would that be better?

Reply:
I've illustrated that change in this commit: https://github.com/kubernetes/perf-tests/pull/1694/commits/84669a2a47b255ad702a31b8f323420da5b9f95f Let me know if I'm still missing something here 🤔